server/src/db_manager.py

```
@@ -0,0 +1,197 @@
+            except UnboundLocalError:
+                pass
+
+    def get_bill_from_id(self, id):
+        """returns tuple (bill_id, bill_name, account_username, timestamp, total, is_paid)"""
+        conn = sqlite3.connect(self.db_name)
+        c = conn.cursor()
+        try:
+            c.execute("""select * from bills where bill_id=?""", (id,))
+            bill = c.fetchone()
+            return bill
+        except Exception as e:
+            print(e)
+        finally:
+            try:
+                conn.close()
+            except UnboundLocalError:
+                pass
+
+    def get_payments_from_id(self, id):
+        """returns tuple (bill_id, payee_name, amount_owed, is_paid)"""
+        conn = sqlite3.connect(self.db_name)
+        c = conn.cursor()
+        try:
+            c.execute("""select * from payments where bill_id=?""", (id,))
+            payments = c.fetchall()
+            return payments
+        except Exception as e:
+            print(e)
+        finally:
+            try:
+                conn.close()
+            except UnboundLocalError:
+                pass
+
+    def get_bills_from_username(self, username):
+        """returns list of tuple (bill_id, bill_name, account_username, timestamp, total, is_paid)"""
+        conn = sqlite3.connect(self.db_name)
+        c = conn.cursor()
+        try:
+            c.execute("""select * from bills where username=?""", (username,))
+            bills = c.fetchall()
+            return bills
+        except Exception as e:
+            print(e)
+        finally:
+            try:
+                conn.close()
+            except UnboundLocalError:
+                pass
+
+
+    def add_user(self, username, password):
+        conn = sqlite3.connect(self.db_name)
+        c = conn.cursor()
+        try:
+            c.execute("""insert into users
+                    (username, password) 
+                    values 
+                    (?, ?)""", (username, password))
```

comment: What happens when someone attempts to add a repeated username?

---

User field has constraint "on conflict ignore" so any attempt to add duplicate users will fail silently.

Ignore might not be ideal because register should return the result. Maybe abort?